### ISSUE TITLE
Restrict rules to a minimum parserOptions.ecmaVersion

### ISSUE THREAD

**Issue by Moeriki** [eslint/eslint#5968]
**Problem**

Right now half of my projects are written in ES6. Half in ES5. I have a shareable config I re-use between my projects. The problem I have is that I have to create separate sharable configs for my ES5/ES6 projects. Because rules such as `prefer-template` or `prefer-spread` give warnings regardless of my configured `parserOptions.ecmaVersion`.

**Solution**

A minimum ecmaVersion configurable per rule.

```javascript
module.exports = {
    meta: {
        ecmaVersion: 6
    }
}
```

**Note**: I know you can create multiple files in a sharable configs and extend those fitting the specific project. This is how I solve this right now.

I'd like to know if this would be a good idea or not.

**Comment by platinumazure** [comment]
This is interesting but needs to be fleshed out more. What should happen if I am using a rule with ecmaVersion: 6 but my parser options are set to ecmaVersion: 5? How would this work with custom parsers which may have a completely different options interface?

**Comment by zxqfox** [comment]
Some naive answers:

> What (exactly) should happen if I am using a rule with ecmaVersion: 6 but my parser options are set to ecmaVersion: 5?

Since rule requires ecma 6, it can't be called on the code. So rule should be just skipped.

> How would this work with custom parsers which may have a completely different options interface and may not support ecmaVersion whatsoever?

Hmm. Why it should be an option for parser, since it's an option for eslint itself to filter out rules or whatever? If option wasn't passed explicitly we can use the most common by default (ES6 for now?).

**Comment by platinumazure** [comment]
> Hmm. Why it should be an option for parser, since it's an option for eslint itself to filter out rules or whatever? If option wasn't passed explicitly we can use the most common by default (ES6 for now?).

Well, at the present moment, `ecmaVersion` is specified under `parserOptions` in config. The whole `parserOptions` block is sent to the specified parser (or espree by default). Some parsers might not understand `ecmaVersion`. And I don't think it's a good idea to overload the responsibility of that particular parser option, because then the configuration design violates the Single Responsibility Principle and could result in users writing in with issues due to unexpected rule ignores (or the parser not liking their options, or...).

I (personally, not speaking for the team) could get behind *a* `ecmaVersion` rule constraint, as long as it's not *the same* `ecmaVersion` as what is used in parser options. But even then, it would probably need a different nomenclature to reduce chance of confusion.

As it is, it's possible to warn within rules for this:

```js
if (context.parserOptions.ecmaVersion < 5) {
    context.report({
        node: node,
        message: "Cannot run rule on this file: ecmaVersion must be greater than 5"
    });
}
```

But even that still presupposes a particular parser configuration.

**Comment by nzakas** [comment]
What @platinumazure said. The options aren't really for ESLint, they are for the parser (that's why it's named `parserOptions`) and there's no guarantee that a parser will implement `ecmaVersion`, so there's only a loose connection between the parser and the rules. I'm not sure we would want to introduce a tight coupling there when we have no real guarantee. (It also doesn't make much sense to enable a rule in your config and have it not actually do anything due to some implicit check.)

What others have done is included both an ES5 and an ES6 config in their shareable config, like Airbnb: https://www.npmjs.com/package/eslint-config-airbnb-base

**Comment by nzakas** [comment]
Per [TSC meeting](https://github.com/eslint/tsc-meetings/blob/master/notes/2016/07-Jul-2016.md#httpsgithubcomeslinteslintissues5968), we don't believe that it makes sense to disable ES6-only rules when `ecmaVersion` is less than 6. We don't think the user experience of enabling a rule and having it disable itself based on `ecmaVersion` is good or expected and instead recommend people create separate ES5 and ES6 ESLint configs when necessary.

**Comment by platinumazure** [comment]
Just to add, glob-based configuration (if that ever gets implemented) could make it easier to maintain one configuration that allows some files to use es6 rules.
